File: packages/cogentlm/native/runtime/scheduler/slot_scheduler.cpp
/////////////////////////////////////////////////////////////////////////////////////////////////
//
// slot_scheduler.cpp
//
// - Single-threaded scheduler.
//
/////////////////////////////////////////////////////////////////////////////////////////////////

#include "runtime/scheduler/slot_scheduler.h"

#include <algorithm>
#include <chrono>
#include <utility>

namespace {

double duration_ms(std::chrono::steady_clock::time_point start,
                   std::chrono::steady_clock::time_point end) {
  return std::chrono::duration<double, std::milli>(end - start).count();
}

} // namespace

namespace noumena::cogentengine {

void SlotState::ResetToIdle() {
  if (sampler != nullptr) {
    llama_sampler_free(sampler);
  }
  phase = SlotPhase::Idle;
  seq_id = -1;
  request_id = 0;
  request = nullptr;
  session = nullptr;
  prefill_cursor = 0;
  decode_step_count = 0;
  batch_participation_count = 0;
  generated_tokens.clear();
  output_text.clear();
  buffered_output_text.clear();
  pending_utf8_bytes.clear();
  terminal_error_message.clear();
  sampler = nullptr;
  mirror.current_kv_tokens.clear();
  mirror.n_past = 0;
  mirror.hardware_id = -1;
}

void SlotState::AttachRequest(GenerateRequest &request_ref,
                              SequenceState &session_ref) {
  request_id = request_ref.id;
  request = &request_ref;
  session = &session_ref;
  mirror.current_kv_tokens = session_ref.current_kv_tokens;
  mirror.n_past = session_ref.n_past;
  mirror.hardware_id = session_ref.hardware_id;
  phase = SlotPhase::Admitted;
  prefill_cursor = 0;
  decode_step_count = 0;
  batch_participation_count = 0;
  generated_tokens.clear();
  output_text.clear();
  buffered_output_text.clear();
  pending_utf8_bytes.clear();
  terminal_error_message.clear();
}

void SlotScheduler::Resize(std::size_t slot_count) {
  slots_.resize(slot_count);
  for (std::size_t i = 0; i < slots_.size(); ++i) {
    slots_[i].slot_id = i;
    slots_[i].seq_id = -1;
    if (slots_[i].phase == SlotPhase::Idle && slots_[i].request == nullptr) {
      continue;
    }
    slots_[i].ResetToIdle();
    slots_[i].slot_id = i;
    slots_[i].seq_id = -1;
  }
}

SlotState *SlotScheduler::FindFirstActiveSlot() {
  auto active_slot_it =
      std::find_if(slots_.begin(), slots_.end(), [](const SlotState &slot) {
        return slot.request != nullptr && slot.phase != SlotPhase::Idle &&
               slot.phase != SlotPhase::Completed &&
               slot.phase != SlotPhase::Failed;
      });
  return active_slot_it == slots_.end() ? nullptr : &(*active_slot_it);
}

void SlotScheduler::SelectDecodeReadySlots(
    std::vector<SlotState *> &out_slots) {
  out_slots.clear();
  out_slots.reserve(slots_.size());

  // Phase 4 algorithm steps:
  // 1. Admit only slots that are already decode-ready for this tick.
  // 2. Exclude slots that still have buffered text waiting to be emitted,
  //    because those slots should not consume additional decode budget yet.
  // 3. Keep ordering deterministic by preserving slot order.
  for (SlotState &slot : slots_) {
    if (slot.request == nullptr || slot.session == nullptr) {
      continue;
    }
    if (slot.phase != SlotPhase::Decode) {
      continue;
    }
    if (slot.generated_tokens.empty()) {
      continue;
    }
    if (!slot.buffered_output_text.empty()) {
      continue;
    }
    out_slots.push_back(&slot);
  }
}

void SlotScheduler::SelectPrefillReadySlots(
    std::vector<SlotState *> &out_slots) {
  out_slots.clear();
  out_slots.reserve(slots_.size());

  // Phase 4 algorithm steps:
  // 1. Admit only slots that still have prompt tokens left to prefill.
  // 2. Keep selection free of fairness heuristics; chunking and reservation
  //    policy belong in the tick budget and batch planner.
  // 3. Preserve slot order so later policy behavior is explainable.
  for (SlotState &slot : slots_) {
    if (slot.request == nullptr || slot.session == nullptr) {
      continue;
    }
    if (slot.phase != SlotPhase::Prefill && slot.phase != SlotPhase::Admitted) {
      continue;
    }
    if (slot.request->is_multimodal_turn &&
        slot.request->multimodal.has_value()) {
      out_slots.push_back(&slot);
      continue;
    }
    if (slot.prefill_cursor >= slot.request->prompt_tokens.size()) {
      continue;
    }
    out_slots.push_back(&slot);
  }
}

SchedulerTickBudget
SlotScheduler::BuildTickBudget(const SchedulerPolicyConfig &policy,
                               int32_t decode_ready_count,
                               int32_t prefill_ready_count,
                               int32_t max_batch_tokens,
                               int32_t prefill_chunk_size) {
  (void)prefill_chunk_size;
  SchedulerTickBudget budget;
  budget.total_token_budget = std::max(0, max_batch_tokens);
  budget.decode_first = decode_ready_count > 0;

  if (budget.total_token_budget <= 0) {
    return budget;
  }

  const int32_t clamped_decode_ready =
      std::max<int32_t>(0, decode_ready_count);
  const int32_t clamped_prefill_ready =
      std::max<int32_t>(0, prefill_ready_count);

  if (clamped_decode_ready == 0) {
    budget.reserved_decode_tokens = 0;
    budget.reserved_prefill_tokens = budget.total_token_budget;
    return budget;
  }

  if (clamped_prefill_ready == 0) {
    budget.reserved_decode_tokens =
        std::min(clamped_decode_ready, budget.total_token_budget);
    budget.reserved_prefill_tokens =
        budget.total_token_budget - budget.reserved_decode_tokens;
    return budget;
  }

  const int32_t requested_decode_reserve =
      policy.decode_token_reserve > 0
          ? std::min(policy.decode_token_reserve, clamped_decode_ready)
          : clamped_decode_ready;
  const int32_t decode_ready_budget =
      std::min(clamped_decode_ready, budget.total_token_budget);

  switch (policy.mode) {
  case SchedulerPolicyMode::LatencyFirst:
    // Decode latency wins. Prefill uses leftover capacity only.
    budget.reserved_decode_tokens =
        policy.decode_token_reserve > 0
            ? std::min(decode_ready_budget, requested_decode_reserve)
            : decode_ready_budget;
    break;
  case SchedulerPolicyMode::ThroughputFirst: {
    // Keep decode alive, but bias the shared batch toward prompt work.
    const int32_t prefill_floor =
        budget.total_token_budget > 1
            ? std::max<int32_t>(1, (budget.total_token_budget * 3) / 4)
            : 0;
    const int32_t decode_ceiling =
        std::max<int32_t>(1, budget.total_token_budget - prefill_floor);
    const int32_t throughput_reserve =
        policy.decode_token_reserve > 0 ? requested_decode_reserve : 1;
    budget.reserved_decode_tokens =
        std::min({decode_ready_budget, decode_ceiling, throughput_reserve});
    break;
  }
  case SchedulerPolicyMode::Balanced:
  default: {
    // Preserve decode responsiveness while leaving room for prefill whenever
    // the batch has more than one token of capacity. With n_batch=1, decode
    // must not be starved by the prefill floor.
    const int32_t prefill_floor = budget.total_token_budget > 1 ? 1 : 0;
    const int32_t decode_ceiling =
        std::max(0, budget.total_token_budget - prefill_floor);
    budget.reserved_decode_tokens =
        std::min(decode_ready_budget, decode_ceiling);
    if (policy.decode_token_reserve > 0) {
      budget.reserved_decode_tokens =
          std::min(budget.reserved_decode_tokens, requested_decode_reserve);
    }
    break;
  }
  }

  budget.reserved_prefill_tokens =
      std::max(0, budget.total_token_budget - budget.reserved_decode_tokens);

  return budget;
}

bool SlotScheduler::AdmitPendingRequests(RequestQueue &request_queue,
                                         SessionStore &session_store) {
  auto idle_slot_it =
      std::find_if(slots_.begin(), slots_.end(), [](const SlotState &slot) {
        return slot.phase == SlotPhase::Idle && slot.request == nullptr;
      });
  if (idle_slot_it == slots_.end()) {
    return false;
  }

  const std::optional<GenerateRequestId> next_request_id =
      request_queue.TryPopNextAdmissible(
          [&session_store](const GenerateRequest &request) {
            return session_store.CanAdmit(request.context_key);
          });
  if (!next_request_id.has_value()) {
    return false;
  }

  GenerateRequest *request = request_queue.FindMutable(*next_request_id);
  if (request == nullptr) {
    return false;
  }

  SequenceState &session = session_store.GetOrCreateSession(request->context_key);
  
  const llama_seq_id leased_seq_id = session_store.AcquireSeqId(session.hardware_id);
  if (leased_seq_id < 0) {
    GenerateResponse response;
    response.request_id = request->id;
    response.status = GenerateResponseStatus::Failed;
    response.error_message = "Failed to acquire a hardware sequence ID.";
    request_queue.MarkCompleted(std::move(response));
    return false;
  }

  // LEASE ID: We honor the session's sticky ID if possible to enable LCP reuse.
  // If we get a different ID, we must clear the logical cache as the physical 
  // state will not match.
  if (leased_seq_id != session.hardware_id) {
    session.current_kv_tokens.clear();
    session.n_past = 0;
  }
  session.hardware_id = leased_seq_id;

  session_store.Pin(request->context_key);
  idle_slot_it->AttachRequest(*request, session);
  idle_slot_it->seq_id = leased_seq_id;
  idle_slot_it->phase = SlotPhase::Prefill;
  return true;
}

void SlotScheduler::FinalizeCompletedSlots(RequestQueue &request_queue,
                                          SessionStore &session_store) {
  for (SlotState &slot : slots_) {
    if (slot.phase != SlotPhase::Completed && slot.phase != SlotPhase::Failed) {
      continue;
    }

    GenerateResponse response;
    response.request_id = slot.request_id;
    response.status =
        slot.request != nullptr && slot.request->cancel_requested
            ? GenerateResponseStatus::Cancelled
            : (slot.phase == SlotPhase::Completed
                   ? GenerateResponseStatus::Completed
                   : GenerateResponseStatus::Failed);
    response.output_text = std::move(slot.output_text);
    if (slot.request != nullptr) {
      GenerateRequest &request = *slot.request;
      request.completed_at = std::chrono::steady_clock::now();
      request.has_completed_at = true;

      response.runtime_observability.ttft_ms =
          request.has_first_token_at
              ? duration_ms(request.enqueued_at, request.first_token_at)
              : 0.0;
      response.runtime_observability.itl_avg_ms =
          request.output_tokens > 1
              ? request.itl_sum_ms /
                    static_cast<double>(request.output_tokens - 1)
              : 0.0;
      response.runtime_observability.itl_p99_ms = request.itl_p99_ms;
      response.runtime_observability.e2e_ms =
          duration_ms(request.enqueued_at, request.completed_at);

      response.runtime_observability.prefill_ms = request.prefill_ms;
      response.runtime_observability.decode_ms = request.decode_ms;

      response.runtime_observability.native_gpu_ms = request.native_gpu_ms;
      response.runtime_observability.native_sync_ms = request.native_sync_ms;
      response.runtime_observability.native_logic_ms = request.native_logic_ms;

      response.runtime_observability.input_tokens =
          request.input_tokens > 0 ? request.input_tokens : static_cast<int32_t>(request.prompt_tokens.size());
      response.runtime_observability.output_tokens = request.output_tokens;
      response.runtime_observability.cache_hits = request.cache_hits;
    }
    if (response.status == GenerateResponseStatus::Cancelled) {
      response.error_message = "Request cancelled.";
    } else if (slot.phase == SlotPhase::Failed) {
      response.error_message =
          slot.terminal_error_message.empty() ? "Request failed."
                                              : slot.terminal_error_message;
    }

    if (slot.request != nullptr && slot.session != nullptr) {
      slot.session->current_kv_tokens = slot.mirror.current_kv_tokens;
      slot.session->n_past = slot.mirror.n_past;
      slot.session->hardware_id = slot.mirror.hardware_id;
    }

    if (slot.request != nullptr) {
      const std::string context_key = slot.request->context_key;
      const bool drop_multimodal_session = slot.request->is_multimodal_turn;
      session_store.Unpin(context_key);
      if (drop_multimodal_session) {
        session_store.Remove(context_key);
      }
    }
    
    if (slot.seq_id >= 0) {
      session_store.ReleaseSeqId(slot.seq_id);
      slot.seq_id = -1;
    }
    
    request_queue.MarkCompleted(std::move(response));
    slot.ResetToIdle();
  }
}

void SlotScheduler::EmitBufferedTokenPiece(RequestQueue &request_queue,
                                          SlotState &slot) {
  if (slot.buffered_output_text.empty()) {
    return;
  }
  GenerateRequest *request = slot.request;

  if (request != nullptr) {
    const auto now = std::chrono::steady_clock::now();
    if (request->has_last_token_at) {
      const double itl_ms = duration_ms(request->last_token_at, now);
      request->itl_sum_ms += itl_ms;
      request->itl_p99_ms = std::max(request->itl_p99_ms, itl_ms);
    }
    request->last_token_at = now;
    request->has_last_token_at = true;

    if (request->token_emission_mode ==
        GenerateTokenEmissionMode::StreamingBuffer) {
      request_queue.AppendStreamingToken(request->id,
                                         slot.buffered_output_text);
    }
  }

  slot.output_text.append(slot.buffered_output_text);
  slot.buffered_output_text.clear();
}

} // namespace noumena::cogentengine

File: packages/cogentlm-rs/crates/cogentlm-core/src/runtime/scheduler/slot_scheduler.rs
use std::time::{Duration, Instant};

use crate::runtime::config::{SchedulerPolicyConfig, SchedulerPolicyMode, SchedulerTickBudget};
use crate::runtime::metrics::RuntimeObservabilityMetrics;
use crate::runtime::request::{
    GenerateRequest, GenerateResponse, GenerateResponseStatus, GenerateTokenEmissionMode,
    RequestQueue,
};
use crate::runtime::session::SessionStore;

use super::{SlotPhase, SlotState};

#[derive(Debug, Default)]
pub struct SlotScheduler {
    slots: Vec<SlotState>,
}

impl SlotScheduler {
    pub fn resize(&mut self, slot_count: usize) {
        if slot_count < self.slots.len() {
            for slot in &mut self.slots[slot_count..] {
                slot.reset_to_idle();
            }
        }

        self.slots.resize_with(slot_count, SlotState::default);
        for (slot_id, slot) in self.slots.iter_mut().enumerate() {
            slot.slot_id = slot_id;
            slot.seq_id = -1;
            if slot.phase == SlotPhase::Idle && slot.request().is_none() {
                continue;
            }
            slot.reset_to_idle();
            slot.slot_id = slot_id;
            slot.seq_id = -1;
        }
    }

    pub fn find_first_active_slot(&self) -> Option<usize> {
        self.slots.iter().position(|slot| {
            slot.request().is_some()
                && slot.phase != SlotPhase::Idle
                && slot.phase != SlotPhase::Completed
                && slot.phase != SlotPhase::Failed
        })
    }

    pub fn select_decode_ready_slots(&self) -> Vec<usize> {
        let mut out = Vec::new();
        self.select_decode_ready_slots_into(&mut out);
        out
    }

    /// Fill `out` with the indices of slots ready to decode. Clears the
    /// caller's buffer first so it can be reused across ticks without
    /// reallocating each time.
    pub fn select_decode_ready_slots_into(&self, out: &mut Vec<usize>) {
        out.clear();
        for (index, slot) in self.slots.iter().enumerate() {
            let request_ready = slot.request().is_some() && slot.session.is_some();
            let slot_ready = slot.phase == SlotPhase::Decode
                && !slot.generated_tokens.is_empty()
                && slot.buffered_output_text.is_empty();
            if request_ready && slot_ready {
                out.push(index);
            }
        }
    }

    pub fn select_prefill_ready_slots(&self) -> Vec<usize> {
        let mut out = Vec::new();
        self.select_prefill_ready_slots_into(&mut out);
        out
    }

    /// Fill `out` with the indices of slots ready to prefill. See
    /// [`Self::select_decode_ready_slots_into`].
    pub fn select_prefill_ready_slots_into(&self, out: &mut Vec<usize>) {
        out.clear();
        for (index, slot) in self.slots.iter().enumerate() {
            let Some(request) = slot.request() else {
                continue;
            };
            if slot.session.is_none() {
                continue;
            }
            if slot.phase != SlotPhase::Prefill && slot.phase != SlotPhase::Admitted {
                continue;
            }
            if request.is_multimodal_turn && request.multimodal.is_some() {
                out.push(index);
                continue;
            }
            if slot.prefill_cursor < request.prompt_tokens.len() {
                out.push(index);
            }
        }
    }

    pub fn slots(&self) -> &[SlotState] {
        &self.slots
    }

    pub fn mutable_slots(&mut self) -> &mut [SlotState] {
        &mut self.slots
    }

    pub fn build_tick_budget(
        policy: SchedulerPolicyConfig,
        decode_ready_count: i32,
        prefill_ready_count: i32,
        max_batch_tokens: i32,
        _prefill_chunk_size: i32,
    ) -> SchedulerTickBudget {
        let mut budget = SchedulerTickBudget {
            total_token_budget: max_batch_tokens.max(0),
            decode_first: decode_ready_count > 0,
            ..SchedulerTickBudget::default()
        };

        if budget.total_token_budget <= 0 {
            return budget;
        }

        let clamped_decode_ready = decode_ready_count.max(0);
        let clamped_prefill_ready = prefill_ready_count.max(0);

        if clamped_decode_ready == 0 {
            budget.reserved_decode_tokens = 0;
            budget.reserved_prefill_tokens = budget.total_token_budget;
            return budget;
        }

        if clamped_prefill_ready == 0 {
            budget.reserved_decode_tokens = clamped_decode_ready.min(budget.total_token_budget);
            budget.reserved_prefill_tokens =
                budget.total_token_budget - budget.reserved_decode_tokens;
            return budget;
        }

        let requested_decode_reserve = if policy.decode_token_reserve > 0 {
            policy.decode_token_reserve.min(clamped_decode_ready)
        } else {
            clamped_decode_ready
        };
        let decode_ready_budget = clamped_decode_ready.min(budget.total_token_budget);

        budget.reserved_decode_tokens = match policy.mode {
            SchedulerPolicyMode::LatencyFirst => {
                if policy.decode_token_reserve > 0 {
                    decode_ready_budget.min(requested_decode_reserve)
                } else {
                    decode_ready_budget
                }
            }
            SchedulerPolicyMode::ThroughputFirst => {
                let prefill_floor = if budget.total_token_budget > 1 {
                    ((budget.total_token_budget * 3) / 4).max(1)
                } else {
                    0
                };
                let decode_ceiling = (budget.total_token_budget - prefill_floor).max(1);
                let throughput_reserve = if policy.decode_token_reserve > 0 {
                    requested_decode_reserve
                } else {
                    1
                };
                decode_ready_budget
                    .min(decode_ceiling)
                    .min(throughput_reserve)
            }
            SchedulerPolicyMode::Balanced => {
                let prefill_floor = if budget.total_token_budget > 1 { 1 } else { 0 };
                let decode_ceiling = (budget.total_token_budget - prefill_floor).max(0);
                let mut decode_tokens = decode_ready_budget.min(decode_ceiling);
                if policy.decode_token_reserve > 0 {
                    decode_tokens = decode_tokens.min(requested_decode_reserve);
                }
                decode_tokens
            }
        };

        budget.reserved_prefill_tokens =
            (budget.total_token_budget - budget.reserved_decode_tokens).max(0);
        budget
    }

    pub fn is_idle(&self) -> bool {
        self.slots.iter().all(|slot| slot.phase == SlotPhase::Idle)
    }

    pub fn admit_pending_requests(
        &mut self,
        request_queue: &mut RequestQueue,
        session_store: &mut SessionStore,
    ) -> bool {
        let debug_metrics_admit_start = Instant::now();
        let Some(idle_slot_index) = self
            .slots
            .iter()
            .position(|slot| slot.phase == SlotPhase::Idle && slot.request().is_none())
        else {
            return false;
        };

        let Some(next_request_id) = request_queue
            .try_pop_next_admissible(|request| session_store.can_admit(&request.context_key))
        else {
            return false;
        };

        let Some(mut request) = request_queue.find(next_request_id).cloned() else {
            return false;
        };

        let context_key = request.context_key.clone();
        let sticky_hardware_id = {
            let session = session_store.get_or_create_session(context_key.clone());
            session.hardware_id
        };
        let leased_seq_id = session_store.acquire_seq_id(sticky_hardware_id);
        if leased_seq_id < 0 {
            request_queue.mark_completed(GenerateResponse {
                request_id: request.id,
                status: GenerateResponseStatus::Failed,
                error_message: "Failed to acquire a hardware sequence ID.".to_string(),
                ..GenerateResponse::default()
            });
            return false;
        }

        let session_snapshot = {
            let session = session_store
                .find_mut(&context_key)
                .expect("session exists after get_or_create_session");
            if leased_seq_id != session.hardware_id {
                session.current_kv_tokens.clear();
                session.n_past = 0;
            }
            session.hardware_id = leased_seq_id;

            // Move out tokens, clone cheap metadata fields, then move tokens into snapshot
            let tokens = std::mem::take(&mut session.current_kv_tokens);
            let mut snap = session.clone();
            snap.current_kv_tokens = tokens;
            snap
        };

        session_store.pin(&context_key);
        let slot = &mut self.slots[idle_slot_index];
        request.debug_metrics_admit_ms += duration_ms(debug_metrics_admit_start, Instant::now());
        slot.attach_request(request, session_snapshot);
        slot.seq_id = leased_seq_id;
        slot.phase = SlotPhase::Prefill;
        true
    }

    pub fn finalize_completed_slots(
        &mut self,
        request_queue: &mut RequestQueue,
        session_store: &mut SessionStore,
    ) {
        for slot in &mut self.slots {
            if slot.phase != SlotPhase::Completed && slot.phase != SlotPhase::Failed {
                continue;
            }

            let debug_metrics_finalize_start = Instant::now();
            let request_snapshot = slot.request().cloned();
            let queue_cancel_requested = request_queue
                .find(slot.request_id)
                .is_some_and(|request| request.cancel_requested);
            let mut metrics_request: Option<(GenerateRequest, Instant)> = None;

            let mut response = GenerateResponse {
                request_id: slot.request_id,
                status: if queue_cancel_requested
                    || request_snapshot
                        .as_ref()
                        .is_some_and(|request| request.cancel_requested)
                {
                    GenerateResponseStatus::Cancelled
                } else if slot.phase == SlotPhase::Completed {
                    GenerateResponseStatus::Completed
                } else {
                    GenerateResponseStatus::Failed
                },
                output_text: std::mem::take(&mut slot.output_text),
                ..GenerateResponse::default()
            };

            if let Some(mut request) = request_snapshot {
                let completed_at = Instant::now();
                request.completed_at = Some(completed_at);
                request.output_tokens = slot.generated_tokens.len().min(i32::MAX as usize) as i32;
                request.emitted_token_count = request.output_tokens;

                if let Some(session) = session_store.find_mut(&request.context_key) {
                    // The slot is about to be reset, so take the token Vec
                    // by-move instead of cloning. For long contexts this
                    // saves an O(N) i32 copy on every request finalisation.
                    session.current_kv_tokens = std::mem::take(&mut slot.mirror.current_kv_tokens);
                    session.n_past = slot.mirror.n_past;
                    session.hardware_id = slot.mirror.hardware_id;
                }

                session_store.unpin(&request.context_key);
                if request.is_multimodal_turn {
                    session_store.remove(&request.context_key);
                }
                metrics_request = Some((request, completed_at));
            }

            if response.status == GenerateResponseStatus::Cancelled {
                response.error_message = "Request cancelled.".to_string();
            } else if slot.phase == SlotPhase::Failed {
                response.error_message = if slot.terminal_error_message.is_empty() {
                    "Request failed.".to_string()
                } else {
                    slot.terminal_error_message.clone()
                };
            }

            if slot.seq_id >= 0 {
                session_store.release_seq_id(slot.seq_id);
                slot.seq_id = -1;
            }

            if let Some((mut request, completed_at)) = metrics_request {
                request.debug_metrics_finalize_ms +=
                    duration_ms(debug_metrics_finalize_start, Instant::now());
                response.runtime_observability = metrics_from_request(&request, completed_at);
            }

            request_queue.mark_completed(response);
            slot.reset_to_idle();
        }
    }

    pub fn emit_buffered_token_piece(request_queue: &mut RequestQueue, slot: &mut SlotState) {
        if slot.buffered_output_text.is_empty() {
            return;
        }

        let buffered = std::mem::take(&mut slot.buffered_output_text);
        let request_id = slot.request_id;
        let mut should_emit = false;

        if let Some(request) = slot.request_mut() {
            should_emit = request.token_emission_mode == GenerateTokenEmissionMode::TokenStream;
        }

        if should_emit {
            request_queue.append_streaming_token(request_id, &buffered);
        }
        slot.output_text.push_str(&buffered);
    }
}

fn metrics_from_request(
    request: &GenerateRequest,
    completed_at: Instant,
) -> RuntimeObservabilityMetrics {
    RuntimeObservabilityMetrics {
        ttft_ms: request
            .first_token_at
            .and_then(|first_token_at| {
                request
                    .enqueued_at
                    .map(|enqueued| duration_ms(enqueued, first_token_at))
            })
            .unwrap_or(0.0),
        itl_avg_ms: average_inter_token_ms(request.output_tokens, request.decode_ms),
        itl_p99_ms: request.itl_p99_ms,
        e2e_ms: request
            .enqueued_at
            .map(|enqueued| duration_ms(enqueued, completed_at))
            .unwrap_or(0.0),
        prefill_ms: request.prefill_ms,
        decode_ms: request.decode_ms,
        native_gpu_ms: request.native_gpu_ms,
        native_sync_ms: request.native_sync_ms,
        native_logic_ms: request.native_logic_ms,
        input_tokens: if request.input_tokens > 0 {
            request.input_tokens
        } else {
            request.prompt_tokens.len() as i32
        },
        output_tokens: request.output_tokens,
        cache_hits: request.cache_hits,
        prefill_tokens: request.prefill_tokens,
        debug_metrics_scheduler_ticks: request.debug_metrics_scheduler_ticks,
        debug_metrics_decode_ticks: request.debug_metrics_decode_ticks,
        debug_metrics_prefill_ticks: request.debug_metrics_prefill_ticks,
        debug_metrics_backend_sampler_attach_attempts: request
            .debug_metrics_backend_sampler_attach_attempts,
        debug_metrics_backend_sampler_attach_failures: request
            .debug_metrics_backend_sampler_attach_failures,
        debug_metrics_admit_ms: request.debug_metrics_admit_ms,
        debug_metrics_normalize_ms: request.debug_metrics_normalize_ms,
        debug_metrics_backend_sampler_attach_ms: request.debug_metrics_backend_sampler_attach_ms,
        debug_metrics_select_slots_ms: request.debug_metrics_select_slots_ms,
        debug_metrics_plan_ms: request.debug_metrics_plan_ms,
        debug_metrics_batch_build_ms: request.debug_metrics_batch_build_ms,
        debug_metrics_llama_decode_ms: request.debug_metrics_llama_decode_ms,
        debug_metrics_llama_sync_ms: request.debug_metrics_llama_sync_ms,
        debug_metrics_apply_bookkeeping_ms: request.debug_metrics_apply_bookkeeping_ms,
        debug_metrics_apply_decode_results_ms: request.debug_metrics_apply_decode_results_ms,
        debug_metrics_sample_ms: request.debug_metrics_sample_ms,
        debug_metrics_token_piece_ms: request.debug_metrics_token_piece_ms,
        debug_metrics_emit_ms: request.debug_metrics_emit_ms,
        debug_metrics_prefix_queue_ms: request.debug_metrics_prefix_queue_ms,
        debug_metrics_finalize_ms: request.debug_metrics_finalize_ms,
        debug_metrics_commit_observability_ms: request.debug_metrics_commit_observability_ms,
        debug_metrics_post_decode_ms: request.debug_metrics_post_decode_ms,
    }
}

fn average_inter_token_ms(output_tokens: i32, decode_ms: f64) -> f64 {
    if output_tokens > 1 {
        decode_ms / f64::from(output_tokens - 1)
    } else {
        0.0
    }
}

fn duration_ms(start: Instant, end: Instant) -> f64 {
    duration_as_ms(end.saturating_duration_since(start))
}

fn duration_as_ms(duration: Duration) -> f64 {
    duration.as_secs_f64() * 1000.0
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::runtime::config::{SchedulerPolicyConfig, SchedulerPolicyMode};
    use crate::runtime::request::GenerateRequestId;
    use crate::runtime::session::SequenceState;

    fn request(id: GenerateRequestId, context_key: &str) -> GenerateRequest {
        let mut request = GenerateRequest::new(id, context_key);
        request.prompt_tokens = vec![1, 2, 3];
        request
    }

    #[test]
    fn resize_resets_non_idle_slots_and_preserves_slot_ids() {
        let mut scheduler = SlotScheduler::default();
        scheduler.resize(2);
        scheduler.slots[0].attach_request(request(1, "a"), SequenceState::default());
        scheduler.slots[0].phase = SlotPhase::Decode;

        scheduler.resize(2);

        assert_eq!(scheduler.slots[0].slot_id, 0);
        assert_eq!(scheduler.slots[0].phase, SlotPhase::Idle);
        assert_eq!(scheduler.slots[1].slot_id, 1);
    }

    #[test]
    fn selects_decode_ready_slots_without_buffered_text() {
        let mut scheduler = SlotScheduler::default();
        scheduler.resize(3);
        scheduler.slots[0].attach_request(request(1, "a"), SequenceState::default());
        scheduler.slots[0].session = Some(SequenceState::default());
        scheduler.slots[0].phase = SlotPhase::Decode;
        scheduler.slots[0].generated_tokens.push(10);
        scheduler.slots[1].attach_request(request(2, "b"), SequenceState::default());
        scheduler.slots[1].session = Some(SequenceState::default());
        scheduler.slots[1].phase = SlotPhase::Decode;
        scheduler.slots[1].generated_tokens.push(11);
        scheduler.slots[1].buffered_output_text = "wait".to_string();

        assert_eq!(scheduler.select_decode_ready_slots(), vec![0]);
    }

    #[test]
    fn build_tick_budget_matches_balanced_policy_with_prefill_floor() {
        let budget = SlotScheduler::build_tick_budget(
            SchedulerPolicyConfig {
                mode: SchedulerPolicyMode::Balanced,
                decode_token_reserve: 8,
                enable_adaptive_prefill_chunking: false,
            },
            8,
            2,
            4,
            0,
        );

        assert_eq!(budget.reserved_decode_tokens, 3);
        assert_eq!(budget.reserved_prefill_tokens, 1);
    }

    #[test]
    fn admit_pending_request_leases_sequence_and_pins_session() {
        let mut scheduler = SlotScheduler::default();
        scheduler.resize(1);
        let mut queue = RequestQueue::new();
        assert!(queue.push(request(1, "ctx")));
        let mut sessions = SessionStore::new(2, 1);

        assert!(scheduler.admit_pending_requests(&mut queue, &mut sessions));

        let slot = &scheduler.slots[0];
        assert_eq!(slot.request_id, 1);
        assert_eq!(slot.seq_id, 0);
        assert_eq!(slot.phase, SlotPhase::Prefill);
        assert_eq!(
            sessions.find("ctx").map(|session| session.pin_count),
            Some(1)
        );
        assert!(!sessions.can_admit("other"));
    }

    #[test]
    fn finalize_completed_slot_writes_response_and_releases_session() {
        let mut scheduler = SlotScheduler::default();
        scheduler.resize(1);
        let mut queue = RequestQueue::new();
        assert!(queue.push(request(1, "ctx")));
        let mut sessions = SessionStore::new(2, 1);
        assert!(scheduler.admit_pending_requests(&mut queue, &mut sessions));

        let slot = &mut scheduler.slots[0];
        slot.phase = SlotPhase::Completed;
        slot.output_text = "done".to_string();
        slot.mirror.current_kv_tokens = vec![1, 2, 3, 4];
        slot.mirror.n_past = 4;

        scheduler.finalize_completed_slots(&mut queue, &mut sessions);

        let response = queue.peek_completed_response(1).expect("response");
        assert_eq!(response.status, GenerateResponseStatus::Completed);
        assert_eq!(response.output_text, "done");
        assert_eq!(
            sessions.find("ctx").map(|session| session.pin_count),
            Some(0)
        );
        assert_eq!(
            sessions
                .find("ctx")
                .map(|session| session.current_kv_tokens.clone()),
            Some(vec![1, 2, 3, 4])
        );
        assert_eq!(scheduler.slots[0].phase, SlotPhase::Idle);
        assert!(sessions.can_admit("other"));
    }

    #[test]
    fn emit_buffered_piece_appends_output_and_stream_frame_when_enabled() {
        let mut queue = RequestQueue::new();
        let (producer, consumer) = crate::runtime::request::token_byte_ring(1024);
        queue.set_token_ring_producer(Some(producer));
        let mut slot = SlotState::new(0);
        let mut request = request(1, "ctx");
        request.token_emission_mode = GenerateTokenEmissionMode::TokenStream;
        slot.attach_request(request, SequenceState::default());
        slot.buffered_output_text = "tok".to_string();

        SlotScheduler::emit_buffered_token_piece(&mut queue, &mut slot);

        assert_eq!(slot.output_text, "tok");
        assert_eq!(queue.total_emitted_token_count(), 1);
        let drain = consumer.drain_available(16, 1024);
        assert_eq!(drain.frames.len(), 1);
        assert_eq!(drain.frames[0].stream_id, 1);
        assert_eq!(drain.frames[0].bytes, b"tok");
    }
}
